File: src/app/src/cli/repl/completion.rs
use std::borrow::Cow;
use std::path::Path;
use std::sync::Arc;
use std::sync::RwLock;

use rustyline::completion::Completer;
use rustyline::completion::Pair;
use rustyline::error::ReadlineError;
use rustyline::highlight::Highlighter;
use rustyline::hint::Hinter;
use rustyline::validate::Validator;

use super::commands::command_arg_completions;
use super::commands::command_short_description;
use super::commands::KNOWN_COMMANDS;
use super::render::DIM;
use super::render::RESET;

pub type CompletionStateRef = Arc<RwLock<CompletionState>>;

#[derive(Default)]
pub struct CompletionState {
    pub models: Vec<String>,
    pub session_ids: Vec<String>,
}

pub struct ReplHelper {
    state: CompletionStateRef,
}

impl ReplHelper {
    pub fn new(state: CompletionStateRef) -> Self {
        Self { state }
    }
}

impl Completer for ReplHelper {
    type Candidate = Pair;

    fn complete(
        &self,
        line: &str,
        pos: usize,
        _ctx: &rustyline::Context<'_>,
    ) -> rustyline::Result<(usize, Vec<Pair>)> {
        let prefix = &line[..pos];

        // Only trigger slash-command completion when the prefix looks like a
        // command typed by hand (`/word`), not a pasted file path that happens
        // to start with `/`.  A valid command prefix is `/` followed by ASCII
        // letters only (no digits, dots, extra slashes, etc.).
        if is_slash_prefix(prefix) && !prefix.contains(' ') {
            let matches: Vec<Pair> = KNOWN_COMMANDS
                .iter()
                .filter(|cmd| cmd.starts_with(prefix))
                .map(|cmd| {
                    let cmd_name = &cmd[1..];
                    let desc = command_short_description(cmd_name).unwrap_or("");
                    if desc.is_empty() {
                        Pair {
                            display: cmd.to_string(),
                            replacement: cmd.to_string(),
                        }
                    } else {
                        Pair {
                            display: format!("{cmd:<12} {desc}"),
                            replacement: cmd.to_string(),
                        }
                    }
                })
                .collect();
            return Ok((0, matches));
        }

        if is_slash_prefix(prefix) {
            if let Some(space_pos) = prefix.find(' ') {
                let cmd = &prefix[..space_pos];
                let arg_part = &prefix[space_pos + 1..];
                if !arg_part.contains(' ') {
                    let state = self.state.read().map_err(|_| {
                        ReadlineError::Io(std::io::Error::other("completion state lock poisoned"))
                    })?;
                    let candidates = command_arg_completions(cmd, arg_part, &state);
                    if !candidates.is_empty() {
                        let pairs = candidates
                            .into_iter()
                            .map(|candidate| Pair {
                                display: candidate.clone(),
                                replacement: candidate,
                            })
                            .collect();
                        return Ok((space_pos + 1, pairs));
                    }
                }
            }
        }

        let word_start = prefix.rfind(char::is_whitespace).map_or(0, |i| i + 1);
        let word = &prefix[word_start..];
        if word.is_empty() {
            return Ok((pos, Vec::new()));
        }

        let matches = complete_file_path(word)
            .into_iter()
            .map(|value| Pair {
                display: value.clone(),
                replacement: value,
            })
            .collect();
        Ok((word_start, matches))
    }
}

/// Custom hint that shows a description alongside the completion text but
/// only inserts the command-name remainder when the user presses right-arrow.
pub struct CommandHint {
    /// The full display string (e.g. `lp  show help`)
    display: String,
    /// The part to insert on accept (e.g. `lp`)
    completion: String,
}

impl rustyline::hint::Hint for CommandHint {
    fn display(&self) -> &str {
        &self.display
    }

    fn completion(&self) -> Option<&str> {
        Some(&self.completion)
    }
}

impl Hinter for ReplHelper {
    type Hint = CommandHint;

    fn hint(&self, line: &str, pos: usize, _ctx: &rustyline::Context<'_>) -> Option<CommandHint> {
        if pos != line.len() || !is_slash_prefix(line) {
            return None;
        }
        let typed = &line[1..];
        if typed.contains(' ') {
            return None;
        }

        // Bare `/` → suggest the most common command
        if typed.is_empty() {
            return Some(CommandHint {
                display: "sessions".to_string(),
                completion: "sessions".to_string(),
            });
        }

        for cmd in KNOWN_COMMANDS {
            let cmd_name = &cmd[1..];
            if cmd_name.starts_with(typed) && cmd_name != typed {
                let rest = &cmd_name[typed.len()..];
                let desc = command_short_description(cmd_name).unwrap_or("");
                let display = if desc.is_empty() {
                    rest.to_string()
                } else {
                    format!("{rest}  {desc}")
                };
                return Some(CommandHint {
                    display,
                    completion: rest.to_string(),
                });
            }
        }
        None
    }
}

impl Highlighter for ReplHelper {
    fn highlight_hint<'h>(&self, hint: &'h str) -> Cow<'h, str> {
        Cow::Owned(format!("{DIM}{hint}{RESET}"))
    }
}

impl Validator for ReplHelper {}
impl rustyline::Helper for ReplHelper {}

/// Returns `true` when `text` looks like a hand-typed slash command prefix:
/// `/` followed by zero or more ASCII lowercase letters (and optionally `!`).
/// Pasted paths like `/some/path.rs` or `:/foo/bar` are rejected because
/// they contain characters that never appear in a valid command name.
pub fn is_slash_prefix(text: &str) -> bool {
    let Some(rest) = text.strip_prefix('/') else {
        return false;
    };
    // After the leading `/`, allow only ASCII letters and `!` (for `/clear!`).
    // A space is fine — it separates the command from its argument — but
    // anything before the first space must be pure command chars.
    // A bare `/` (rest is empty) is also valid — it triggers the full command list.
    let cmd_part = rest.split_once(' ').map_or(rest, |(cmd, _)| cmd);
    cmd_part
        .bytes()
        .all(|b| b.is_ascii_lowercase() || b == b'!')
}

pub fn complete_file_path(partial: &str) -> Vec<String> {
    let path = Path::new(partial);

    let (dir, file_prefix) =
        if partial.ends_with('/') || partial.ends_with(std::path::MAIN_SEPARATOR) {
            (partial.to_string(), String::new())
        } else if let Some(parent) = path.parent() {
            let parent_str = if parent.as_os_str().is_empty() {
                ".".to_string()
            } else {
                parent.to_string_lossy().to_string()
            };
            let file_prefix = path
                .file_name()
                .map(|name| name.to_string_lossy().to_string())
                .unwrap_or_default();
            (parent_str, file_prefix)
        } else {
            (".".to_string(), partial.to_string())
        };

    let entries = match std::fs::read_dir(&dir) {
        Ok(entries) => entries,
        Err(_) => return Vec::new(),
    };

    let dir_prefix = if dir == "." && !partial.contains('/') {
        String::new()
    } else if partial.ends_with('/') || partial.ends_with(std::path::MAIN_SEPARATOR) {
        partial.to_string()
    } else {
        let parent = path.parent().unwrap_or(Path::new(""));
        if parent.as_os_str().is_empty() {
            String::new()
        } else {
            format!("{}/", parent.display())
        }
    };

    let mut matches = Vec::new();
    for entry in entries.flatten() {
        let name = entry.file_name().to_string_lossy().to_string();
        if !name.starts_with(&file_prefix) {
            continue;
        }
        let is_dir = entry
            .file_type()
            .map(|value| value.is_dir())
            .unwrap_or(false);
        let candidate = if is_dir {
            format!("{}{}/", dir_prefix, name)
        } else {
            format!("{}{}", dir_prefix, name)
        };
        matches.push(candidate);
    }
    matches.sort();
    matches
}

File: cli/src/commands/completion.ts
/**
 * Tab completion for the REPL input.
 * Supports slash commands and file paths.
 */

import { readdirSync, statSync } from 'fs'
import { join, dirname, basename } from 'path'
import { COMMANDS, HIDDEN_COMMANDS } from './index.js'

/**
 * Returns true when text looks like a hand-typed slash command prefix:
 * `/` followed by zero or more ASCII lowercase letters.
 * Pasted paths like `/some/path.rs` are rejected.
 */
function isSlashPrefix(text: string): boolean {
  if (!text.startsWith('/')) return false
  const rest = text.slice(1)
  const cmdPart = rest.split(/\s/)[0] ?? ''
  return /^[a-z]*$/.test(cmdPart)
}

export interface CompletionResult {
  /** The completed text to replace the current word */
  replacement: string
  /** All candidates (shown when multiple matches) */
  candidates: string[]
  /** Start index of the word being completed in the line */
  wordStart: number
}

/**
 * Compute tab completion for the current input line and cursor position.
 */
export function complete(line: string, cursorCol: number): CompletionResult | null {
  const beforeCursor = line.slice(0, cursorCol)

  // Slash command completion: input looks like a hand-typed command
  if (isSlashPrefix(beforeCursor)) {
    return completeSlashCommand(beforeCursor)
  }

  // File path completion: word containing /
  const word = currentWord(beforeCursor)
  if (word && (word.includes('/') || word.startsWith('~') || word.startsWith('.'))) {
    return completeFilePath(word, cursorCol)
  }

  return null
}

/**
 * Compute inline ghost hint for the current input.
 * Returns gray text to show after the cursor (Rust REPL style).
 *
 * - Bare `/` → compact list of all visible commands
 * - `/he` → `lp — Show help` (single match: completion + description)
 * - `/skill ` → `[install  list  remove]` (sub-command hints)
 */
export function getGhostHint(line: string, cursorCol: number): string {
  const beforeCursor = line.slice(0, cursorCol)
  const afterCursor = line.slice(cursorCol)

  // Only show hints when cursor is at end of line
  if (afterCursor.trim().length > 0) return ''
  if (!isSlashPrefix(beforeCursor)) return ''

  const parts = beforeCursor.split(/\s+/)
  const cmd = parts[0]!.toLowerCase()

  // Sub-command hints (after space)
  if (parts.length >= 2) {
    const partial = parts.slice(1).join(' ').toLowerCase()
    return getSubCommandHint(cmd, partial)
  }

  // Bare `/` → show all visible command names
  if (cmd === '/') {
    return `  [${COMMANDS.map(c => c.name.slice(1)).join('  ')}]`
  }

  // Only match visible commands for ghost hints
  const matches = COMMANDS.filter(c => c.name.startsWith(cmd))

  if (matches.length === 0) return ''

  // Single match — show completion suffix + description
  if (matches.length === 1) {
    const m = matches[0]!
    const suffix = m.name.slice(cmd.length)
    return `${suffix}  ${m.description}`
  }

  // Multiple matches — show common suffix + candidate names
  const common = commonPrefix(matches.map(m => m.name))
  const suffix = common.slice(cmd.length)
  const names = matches.map(m => m.name).join('  ')
  return suffix ? `${suffix}  [${names}]` : `  [${names}]`
}

function getSubCommandHint(cmd: string, partial: string): string {
  // Resolve the command first
  const resolved = COMMANDS.find(c => c.name === cmd || c.name.startsWith(cmd))
  if (!resolved) return ''

  const subcmds = SUB_COMMANDS[resolved.name] ?? []
  if (subcmds.length === 0) return ''

  if (!partial) {
    return `  [${subcmds.join('  ')}]`
  }

  const matches = subcmds.filter(s => s.startsWith(partial))
  if (matches.length === 0) return ''
  if (matches.length === 1) {
    return matches[0]!.slice(partial.length)
  }
  const common = commonPrefix(matches)
  const suffix = common.slice(partial.length)
  return suffix ? `${suffix}  [${matches.join('  ')}]` : `  [${matches.join('  ')}]`
}

const SUB_COMMANDS: Record<string, string[]> = {
  '/help': COMMANDS.map(c => c.name.slice(1)),
  '/skill': ['install', 'list', 'remove'],
  '/env': ['set', 'del', 'load'],
}

// ---------------------------------------------------------------------------
// Slash command completion
// ---------------------------------------------------------------------------

function completeSlashCommand(input: string): CompletionResult | null {
  const parts = input.split(/\s+/)
  const cmd = parts[0]!.toLowerCase()

  // Only complete the command name itself (first word)
  if (parts.length > 1) return null

  const allCmds = [...COMMANDS, ...HIDDEN_COMMANDS]
  const allNames: string[] = []
  for (const c of allCmds) {
    allNames.push(c.name)
    if (c.aliases) allNames.push(...c.aliases)
  }

  const matches = allNames.filter((n) => n.startsWith(cmd))

  if (matches.length === 0) return null
  if (matches.length === 1) {
    return {
      replacement: matches[0]! + ' ',
      candidates: matches,
      wordStart: 0,
    }
  }

  // Multiple matches — complete common prefix
  const common = commonPrefix(matches)
  return {
    replacement: common,
    candidates: matches,
    wordStart: 0,
  }
}

// ---------------------------------------------------------------------------
// File path completion
// ---------------------------------------------------------------------------

function completeFilePath(word: string, cursorCol: number): CompletionResult | null {
  const wordStart = cursorCol - word.length

  // Expand ~
  const expanded = word.startsWith('~')
    ? (process.env.HOME ?? '') + word.slice(1)
    : word

  let dir: string
  let prefix: string

  try {
    const stat = statSync(expanded)
    if (stat.isDirectory()) {
      dir = expanded.endsWith('/') ? expanded : expanded + '/'
      prefix = ''
    } else {
      return null // exact file, nothing to complete
    }
  } catch {
    // Not a valid path yet — complete the last segment
    dir = dirname(expanded)
    prefix = basename(expanded)
  }

  let entries: string[]
  try {
    entries = readdirSync(dir)
  } catch {
    return null
  }

  const matches = prefix
    ? entries.filter((e) => e.startsWith(prefix))
    : entries.filter((e) => !e.startsWith('.')) // hide dotfiles by default

  if (matches.length === 0) return null

  // Build full paths relative to original word
  const dirPrefix = word.endsWith('/')
    ? word
    : word.slice(0, word.length - prefix.length)

  const fullMatches = matches.map((m) => {
    const fullPath = join(dir, m)
    try {
      if (statSync(fullPath).isDirectory()) return dirPrefix + m + '/'
    } catch { /* ignore */ }
    return dirPrefix + m
  })

  if (fullMatches.length === 1) {
    return {
      replacement: fullMatches[0]!,
      candidates: fullMatches,
      wordStart,
    }
  }

  const common = commonPrefix(fullMatches)
  return {
    replacement: common,
    candidates: fullMatches,
    wordStart,
  }
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function currentWord(beforeCursor: string): string | null {
  // Extract the last whitespace-delimited word
  const match = beforeCursor.match(/(\S+)$/)
  return match ? match[1]! : null
}

function commonPrefix(strings: string[]): string {
  if (strings.length === 0) return ''
  let prefix = strings[0]!
  for (let i = 1; i < strings.length; i++) {
    const s = strings[i]!
    let j = 0
    while (j < prefix.length && j < s.length && prefix[j] === s[j]) j++
    prefix = prefix.slice(0, j)
  }
  return prefix
}
